index.html

```diff
@@ -35,58 +41,150 @@
+          <li>If |options|["{{TCPSocketOptions/keepAliveDelay}}"] is less than one second, throw a {{TypeError}}  and return `null`.
+          <li>If |options|["{{TCPSocketOptions/sendBufferSize}}"] is equal to 0, throw a {{TypeError}}  and return `null`.
+          <li>If |options|["{{TCPSocketOptions/receiveBufferSize}}"] is equal to 0, throw a {{TypeError}}  and return `null`.
+          <li>Perform the following steps [=in parallel=].
+            <ol>
+              <li>
+                Invoke the operating system to open the TCP socket using the given |remoteAddress| and |remotePort| and the 
+                connection parameters (or their defaults) specified in |options|.
+              <li>
+                If this fails for any reason:
+                <ol>
+                  <li>[=Reject=] the {{TCPSocket/[[connectionPromise]]}} with a "{{NetworkError}}" {{DOMException}}.
+                  <li>Detach {{TCPSocket/[[closedPromise]]}}.
+                </ol>
+              <li>
+                On success:
+                <ol>
+                  <li>[=initialize tcp readable stream|Initialize=] {{TCPSocket/[[readable]]}}.
+                  <li>[=initialize tcp writable stream|Initialize=] {{TCPSocket/[[writable]]}}.
+                  <li>Let |conn:TCPSocketConnection| be a new {{TCPSocketConnection}}.
+                  <li>Set |conn|["{{TCPSocketConnection/readable}}"] to [=this=].{{TCPSocket/[[readable]]}}.
+                  <li>Set |conn|["{{TCPSocketConnection/writable}}"] to [=this=].{{TCPSocket/[[writable]]}}.
+                  <li>Populate the remaining fields of |conn| using the information provided by the operating system:
+                      |conn|["{{TCPSocketConnection/remoteAddress}}"], |conn|["{{TCPSocketConnection/remotePort}}"],
+                      |conn|["{{TCPSocketConnection/localAddress}}"] and |conn|["{{TCPSocketConnection/localPort}}"].
+                  <li>Resolve [=this=].{{TCPSocket/[[connectionPromise]]}} with |conn|.
+                </ol>
+            </ol>
+          <li>Return [=this=].
+        </ol>
+
+        <section data-dfn-for="TCPSocketOptions">
+          <h4><dfn>TCPSocketOptions</dfn> dictionary</h4>
+
           <pre class="idl">
-          dictionary SocketOptions {
-            DOMString localAddress;
-            unsigned short localPort;
+            dictionary TCPSocketOptions {
+              unsigned long sendBufferSize;
+              unsigned long receiveBufferSize;
 
-            DOMString remoteAddress;
-            unsigned short remotePort;
+              DOMString localAddress;
+              unsigned short localPort;
 
-            unsigned long sendBufferSize;
-            unsigned long receiveBufferSize;
-          };
+              boolean noDelay;
+              boolean keepAlive;
+              unsigned long keepAliveDelay;
```

Comment:
Can these two fields be merged? I think you can tell when a value isn't set for an unsigned long if you don't set a default value.

Reply:
Do we want to expose system default behavior like that? Is it actually different between operating systems? I have a slight preference for having the specification be opinionated on what the default behavior is unless there's a good reason why different platforms have different defaults.

Reply:
I suspect these might be different for Windows/Linux, for example. One option I see is disabling keepAlive by default and allowing the user to specify it themselves.